## Link
[Diabetes 130-US Hospitals for Years 1999-2008](https://archive.ics.uci.edu/dataset/296/diabetes+130-us+hospitals+for+years+1999-20)

- Before delving into the project, let us have a feel of the data.

In [1]:
import pandas as pd

In [4]:
path =  "../diabetic_data.csv"
df = pd.read_csv(path)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

In [7]:
len(df.columns), df.columns

(50,
 Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
        'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
        'time_in_hospital', 'payer_code', 'medical_specialty',
        'num_lab_procedures', 'num_procedures', 'num_medications',
        'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
        'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
        'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
        'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
        'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
        'tolazamide', 'examide', 'citoglipton', 'insulin',
        'glyburide-metformin', 'glipizide-metformin',
        'glimepiride-pioglitazone', 'metformin-rosiglitazone',
        'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
       dtype='object'))

In [ ]:
len(df)

In [8]:
# distribution of the target 
df['readmitted'].value_counts()

NO     54864
>30    35545
<30    11357
Name: readmitted, dtype: int64

In [11]:
df.corr()

encounter_id  patient_nbr  admission_type_id  \
encounter_id                  1.000000     0.512028          -0.158961   
patient_nbr                   0.512028     1.000000          -0.011128   
admission_type_id            -0.158961    -0.011128           1.000000   
discharge_disposition_id     -0.132876    -0.136814           0.083483   
admission_source_id          -0.112402    -0.032568           0.106654   
time_in_hospital             -0.062221    -0.024092          -0.012500   
num_lab_procedures           -0.026062     0.015946          -0.143713   
num_procedures               -0.014225    -0.015570           0.129888   
num_medications               0.076113     0.020665           0.079535   
number_outpatient             0.103756     0.103379           0.026511   
number_emergency              0.082803     0.062352          -0.019116   
number_inpatient              0.030962     0.012480          -0.038161   
number_diagnoses              0.265149     0.226847          -0.117126   

                          discharge_disposition_id  admission_source_id  \
encounter_id                             -0.132876            -0.112402   
patient_nbr                              -0.136814            -0.032568   
admission_type_id                         0.083483             0.106654   
discharge_disposition_id                  1.000000             0.018193   
admission_source_id                       0.018193             1.000000   
time_in_hospital                          0.162748            -0.006965   
num_lab_procedures                        0.023415             0.048885   
num_procedures                            0.015921            -0.135400   
num_medications                           0.108753            -0.054533   
number_outpatient                        -0.008715             0.027244   
number_emergency                         -0.024471             0.059892   
number_inpatient                          0.020787             0.036314   
number_diagnoses                          0.046891             0.072114   

                          time_in_hospital  num_lab_procedures  \
encounter_id                     -0.062221           -0.026062   
patient_nbr                      -0.024092            0.015946   
admission_type_id                -0.012500           -0.143713   
discharge_disposition_id          0.162748            0.023415   
admission_source_id              -0.006965            0.048885   
time_in_hospital                  1.000000            0.318450   
num_lab_procedures                0.318450            1.000000   
num_procedures                    0.191472            0.058066   
num_medications                   0.466135            0.268161   
number_outpatient                -0.008916           -0.007602   
number_emergency                 -0.009681           -0.002279   
number_inpatient                  0.073623            0.039231   
number_diagnoses                  0.220186            0.152773   

                          num_procedures  num_medications  number_outpatient  \
encounter_id                   -0.014225         0.076113           0.103756   
patient_nbr                    -0.015570         0.020665           0.103379   
admission_type_id               0.129888         0.079535           0.026511   
discharge_disposition_id        0.015921         0.108753          -0.008715   
admission_source_id            -0.135400        -0.054533           0.027244   
time_in_hospital                0.191472         0.466135          -0.008916   
num_lab_procedures              0.058066         0.268161          -0.007602   
num_procedures                  1.000000         0.385767          -0.024819   
num_medications                 0.385767         1.000000           0.045197   
number_outpatient              -0.024819         0.045197           1.000000   
number_emergency               -0.038179         0.013180           0.091459   
number_inpatient               -0.066236         0.06

In [17]:
mapping = {'NO':0,
           '>30':1,
           '<30':2}

In [18]:
df['readmitted']= df['readmitted'].replace(mapping)

In [19]:
df.corr()

encounter_id  patient_nbr  admission_type_id  \
encounter_id                  1.000000     0.512028          -0.158961   
patient_nbr                   0.512028     1.000000          -0.011128   
admission_type_id            -0.158961    -0.011128           1.000000   
discharge_disposition_id     -0.132876    -0.136814           0.083483   
admission_source_id          -0.112402    -0.032568           0.106654   
time_in_hospital             -0.062221    -0.024092          -0.012500   
num_lab_procedures           -0.026062     0.015946          -0.143713   
num_procedures               -0.014225    -0.015570           0.129888   
num_medications               0.076113     0.020665           0.079535   
number_outpatient             0.103756     0.103379           0.026511   
number_emergency              0.082803     0.062352          -0.019116   
number_inpatient              0.030962     0.012480          -0.038161   
number_diagnoses              0.265149     0.226847          -0.117126   
readmitted                   -0.031806     0.057623          -0.008950   

                          discharge_disposition_id  admission_source_id  \
encounter_id                             -0.132876            -0.112402   
patient_nbr                              -0.136814            -0.032568   
admission_type_id                         0.083483             0.106654   
discharge_disposition_id                  1.000000             0.018193   
admission_source_id                       0.018193             1.000000   
time_in_hospital                          0.162748            -0.006965   
num_lab_procedures                        0.023415             0.048885   
num_procedures                            0.015921            -0.135400   
num_medications                           0.108753            -0.054533   
number_outpatient                        -0.008715             0.027244   
number_emergency                         -0.024471             0.059892   
number_inpatient                          0.020787             0.036314   
number_diagnoses                          0.046891             0.072114   
readmitted                                0.012450             0.031816   

                          time_in_hospital  num_lab_procedures  \
encounter_id                     -0.062221           -0.026062   
patient_nbr                      -0.024092            0.015946   
admission_type_id                -0.012500           -0.143713   
discharge_disposition_id          0.162748            0.023415   
admission_source_id              -0.006965            0.048885   
time_in_hospital                  1.000000            0.318450   
num_lab_procedures                0.318450            1.000000   
num_procedures                    0.191472            0.058066   
num_medications                   0.466135            0.268161   
number_outpatient                -0.008916           -0.007602   
number_emergency                 -0.009681           -0.002279   
number_inpatient                  0.073623            0.039231   
number_diagnoses                  0.220186            0.152773   
readmitted                        0.057718            0.037976   

                          num_procedures  num_medications  number_outpatient  \
encounter_id                   -0.014225         0.076113           0.103756   
patient_nbr                    -0.015570         0.020665           0.103379   
admission_type_id               0.129888         0.079535           0.026511   
discharge_disposition_id        0.015921         0.108753          -0.008715   
admission_source_id            -0.135400        -0.054533           0.027244   
time_in_hospital                0.191472         0.466135          -0.008916   
num_lab_procedures              0.058066         0.268161          -0.007602   
num_procedures                  1.000000         0.385767          -0.024819   
num_medications                 0.385767         1.000000           0.045197  

In [15]:
df['readmitted'].value_counts()

NO     54864
>30    35545
<30    11357
Name: readmitted, dtype: int64

In [16]:
df['readmitted'].sample(n=10)

32495      NO
40605     >30
94223      NO
29011      NO
50796      NO
67079      NO
75856      NO
75136      NO
89792      NO
100210     NO
Name: readmitted, dtype: object

- From a cursory glance, the project is an tertiary Classification ML project to determine whether the patient was readmitted or not , and if so if it was after or before 30 days

# 

# Diabetes Readmission Prediction Project
---

## Methodology

-   1. Business Understanding
-   2. Data Understanding
-   3. Data Preparation
-   4. Modeling
-   5. Evaluation
-   6. Deployment
---
## 1. Business Understanding

### Introduction
Diabetes is a chronic medical condition characterized by the body's inability to properly regulate blood sugar levels. It affects millions of people worldwide and, if not well-managed, can lead to severe health complications such as cardiovascular disease, kidney failure, and nerve damage. Effective care and monitoring are crucial, especially during and after hospitalizations, to prevent complications and improve patient outcomes.

### Problem Statement
Hospital readmissions for diabetic patients pose a significant burden to both healthcare systems and patients. The dataset for this project represents ten years (1999-2008) of clinical care at 130 U.S. hospitals and integrated delivery networks, encompassing over 100,000 patient records. Each record details a diabetic patient’s hospital encounter, including information on laboratory tests, medications, and hospital stays lasting up to 14 days.

Despite high-quality evidence showing improved outcomes for diabetic patients through preventive and therapeutic interventions, many patients do not receive consistent care. This inconsistency often leads to poor glycemic control, resulting in complications and costly readmissions. These readmissions increase operational costs for hospitals and impact patient well-being.

The maain goal of this project is to develop a machine learning model that can accurately predict the likelihood of readmission for diabetic patients. By identifying high-risk patients, healthcare providers can implement timely interventions to reduce readmissions and improve patient outcomes.

### Key Stakeholders:
- **Healthcare Providers:** Physicians, nurses, and hospital administrators who can use predictions to enhance patient care.
- **Hospital Management:** Decision-makers responsible for reducing operational costs and optimizing resource allocation.
- **Patients:** Individuals diagnosed with diabetes who benefit from improved care and reduced readmissions.
- **Insurance Companies:** Entities interested in minimizing claim costs by encouraging preventive care.

### Objectives:
- **Model Development:** Build a machine learning model to predict the likelihood of early readmission (within 30 days) for diabetic patients.
- **Feature Analysis:** Identify key factors influencing readmission, such as lab test results, length of stay, and medication usage.
- **Actionable Insights:** Provide data-driven recommendations to healthcare providers for targeted interventions.
- **Patient Support:** Improve diabetic patient care by suggesting personalized health management strategies.

### Success Criteria:
- Achieve a predictive model with high recall to minimize false negatives, ensuring fewer high-risk patients are overlooked.
- Deliver insights that support evidence-based healthcare interventions.
- Evaluate model performance using appropriate metrics such as recall, precision, and F1-score.

---

## 2. Data Understanding

### Data Source
Data source link : [Diabetes 130-US Hospitals for Years 1999-2008](https://archive.ics.uci.edu/dataset/296/diabetes+130-us+hospitals+for+years+1999-20)

The dataset contains information extracted from hospital encounters that meet the following criteria:
- Inpatient encounters (hospital admissions) involving diabetic patients.
- Hospital stays of 1 to 14 days.
- Inclusion of laboratory tests and medication administration.


### Data Characteristics
- **Number of Instances:** 101,766 patient records
- **Number of Features:** 47 attributes
- **Data Types:** Categorical and integer features
- **Sensitive Information:** Includes patient demographics such as age, gender, and race
- **Missing Values:** Present in some key attributes, requiring careful handling

### Key Features
- **Patient demographics** (age, race, gender)
- **Admission details** (admission type, length of stay)
- **Clinical data** (number of lab tests, HbA1c test results, number of medications)
- **Healthcare utilization** (number of outpatient, inpatient, and emergency visits in the previous year)

### Data Limitations
- **Class Imbalance:** The target variable (readmission) is highlt imbalanced, likely requiring resampling or class-weighting techniques.
- **Data Gaps:** Missing values in critical features must be addressed through various strategies like imputation or deletion.
- **Outdated Medical Practices:** The dataset, spanning 1999 to 2008, may reflect older medical tests, treatments, and protocols that have since evolved. This could limit the model's relevance by missing recent advancements in diagnostics, medications, and patient care, affecting its accuracy for current patients.


### Potential Challenges
- **Data Imbalance:** Addressing the underrepresentation of early readmissions in the dataset.
- **Data Preprocessing:** Cleaning and transforming raw data to ensure model readiness.
- **Sensitive Data:** Ensuring ethical handling of demographic information like ethnicity.
